engine/runtime-parser-dsl/src/main/java/org/enso/runtime/parser/dsl/IRChild.java

```diff
@@ -7,9 +7,30 @@
 
 /**
  * Constructor parameter annotated with this annotation will be represented as a child field in the
- * generated super class. Children of IR elements form a tree. A child will be part of the methods
- * traversing the tree, like {@code mapExpression} and {@code children}. The parameter type must be
- * a subtype of {@code org.enso.compiler.ir.IR}.
+ * generated super class, and will have a generated getter method with the same name. Children of IR
+ * elements form a tree. A child will be part of the methods traversing the tree, like {@code
+ * mapExpression} and {@code children}.
+ *
+ * <p>The body of the generated getter method for the field depends on the type of the parameter.
+ * The return type of the getter method is generally the same as the type of the parameter, with
+ * some exceptions documented below.
+ *
+ * <p>The parameter types must be one of the following:
+ *
+ * <ul>
+ *   <li>Subtype of {@code org.enso.compiler.core.IR}
+ *   <li>{@link scala.collection.immutable.List} with type parameter that extends {@code
+ *       org.enso.compiler.core.IR}
+ *   <li>{@link scala.Option} with type parameter that extends {@code org.enso.compiler.core.IR}
+ *   <li>{@code org.enso.persist.Persistance.Reference} with type parameter that extends {@code
+ *       org.enso.compiler.core.IR}
+ * </ul>
+ *
+ * <p>If the parameter type is {@code org.enso.persist.Persistance.Reference<T>}, then the return
+ * type of the getter method is {@code T}. Otherwise, the return type is the same as the parameter
```

Comment:
An alternative approach would be to add a parameter to `@IRChild` to use `Persistance.Reference` internally.  Something like `@IRChild(deferred=true)`. Then it would be up to the annotation processor generator to decide what's the right way to _"implement being deferred (on deserialization)"_ - right now it would use `Persistance.Reference`. However it would be an implementation detail - such usage wouldn't be visible from the generated API.

Reply:
AFAIK, Persistance framework requires that the generated class has a field of type `Persistance.Reference`. As such, there should be `Persistance.Reference` parameter in constructor annotated with `@IRChild`. Originally, I have tried implementing this via `java.util.Supplier<T extends IR>`, but it did not work.